### PR TITLE
[WIP] Optimize RNN performance on CPU by fusing elemwise operations

### PR DESCRIPTION
**Pull request by mingfeima**
This PR aims at improving LSTM and GRU performance by fusing elemwise operations in LSTMCell and GRUCell. Benefits:
1. save memory bandwidth
2. reduce autograd engine deduce

At functional level, this PR applies to all RNN features:
1. no special memory format for RNN weights, completely aligns with current weight format
2. inference and training
3. supports double backward
4. co-works with precompute of `input_gates` (time step fusion on input of weight_ih)
5. supports both `PackedSequence` and `PaddedSequence`
6. supports batch_first, dropout, multi layer, bidirectional

Performance:
Inference on single socket of Xeon Skylake 6148, **1.62x** improvement:
```bash
before:
lstm: layers=1, dir=1, batch_size=1, time_step=15, input_size=250, hidden_size=200: time: 0.73 ms
after:
lstm: layers=1, dir=1, batch_size=1, time_step=15, input_size=250, hidden_size=200: time: 0.45 ms
```
Inference on single core of Xeon Skylake 6148, **1.38x** improvement:
```bash
before:
lstm: layers=1, dir=1, batch_size=1, time_step=15, input_size=250, hidden_size=200: time: 0.77 ms
after:
lstm: layers=1, dir=1, batch_size=1, time_step=15, input_size=250, hidden_size=200: time: 0.56 ms
```
Training on single socket of Xeon Skylake 6148, **3.44x** improvement
```bash
before:
lstm: layers=1, dir=1, batch_size=128, time_step=50, input_size=1024, hidden_size=1024: time: 891.87 ms
after:
lstm: layers=1, dir=1, batch_size=128, time_step=50, input_size=1024, hidden_size=1024: time: 258.98 ms
```
All test results have config jemalloc, otherwise the "before" CPU performance would be even worse.

Benchmark script placed at [repo](https://github.com/mingfeima/op_bench-py), `run.sh rnn.py`. For inference, pick size according to suggestion; For training, pick size from RNN-translation of MLPerf.

### CODE REVIEW

**Comment by VitalyFedyunin** [comment]
@glaringlee please take a look

**Comment by mingfeima** [comment]
Rebased! Please review @glaringlee @VitalyFedyunin @ngimel.

cc @jgong5

**Comment by glaringlee** [comment]
@mingfeima 
Will get this reviewed today or tomorrow!

**Review by glaringlee — CHANGES_REQUESTED** [review]

File: aten/src/ATen/native/RNN.cpp [hunk 1]
```diff
@@ -1457,26 +1465,64 @@ std::tuple<Tensor, Tensor> lstm_cell(
...
-    const Tensor& input_bias,
-    const Tensor& hidden_bias,
-    const Tensor& cx,
-    const Tensor& cy) {
+std::tuple<Tensor, Tensor, Tensor>
+_fused_lstm_cell_cpu(
+    const Tensor& input_gates, const Tensor& hidden_gates,
+    const Tensor& cx) {
+  auto workspace = at::empty_like(input_gates, LEGACY_CONTIGUOUS_MEMORY_FORMAT);
+  auto hy = at::empty_like(cx, LEGACY_CONTIGUOUS_MEMORY_FORMAT);
```
[the remark this answers is not in the document]
Comment (mingfeima) [reply aten/src/ATen/native/RNN.cpp]: Fixed!

File: aten/src/ATen/native/RNN.cpp [hunk 2]
```diff
@@ -1457,26 +1465,64 @@ std::tuple<Tensor, Tensor> lstm_cell(
...
-    const Tensor& hidden_gates,
-    const Tensor& input_bias,
-    const Tensor& hidden_bias,
-    const Tensor& cx,
-    const Tensor& cy) {
+std::tuple<Tensor, Tensor, Tensor>
+_fused_lstm_cell_cpu(
+    const Tensor& input_gates, const Tensor& hidden_gates,
+    const Tensor& cx) {
+  auto workspace = at::empty_like(input_gates, LEGACY_CONTIGUOUS_MEMORY_FORMAT);
```
[the remark this answers is not in the document]
Comment (mingfeima) [reply aten/src/ATen/native/RNN.cpp]: Fixed!

File: aten/src/ATen/native/RNN.cpp [hunk 3]
```diff
@@ -1505,30 +1550,34 @@ _thnn_differentiable_lstm_cell_backward(
   gfg = at::sigmoid_backward(gfg, f);
   gcg = at::tanh_backward(gcg, c);
   Tensor grad_gates = at::cat({gig, gfg, gcg, gog}, 1);
-  Tensor grad_bias = input_bias.defined() ? grad_gates.sum(0, /*keepdim=*/false) : at::Tensor{};
-  return std::make_tuple(grad_gates, grad_gates, std::move(gcx), grad_bias, grad_bias);
+  return std::make_tuple(grad_gates, std::move(gcx));
```
[the remark this answers is not in the document]
Comment (mingfeima) [reply aten/src/ATen/native/RNN.cpp]: yes, right~

**Comment by glaringlee** [comment]
@mingfeima 
It seems your code breaks our asan built (address sanitizer) tests in the following: (potential memory leak due to vectorization)
ci/circleci: pytorch_linux_xenial_py3_clang5_asan_test1 
ci/circleci: pytorch_linux_xenial_py3_clang5_asan_test2

They all call nn.LSTM.

One of failure is this:
https://app.circleci.com/pipelines/github/pytorch/pytorch/194791/workflows/41dd2318-66e5-416b-8f87-e9fb6e8151e5/jobs/6428572
The code is here:
https://github.com/pytorch/pytorch/blob/ca1b8ebbcb8f959c4044d5832acb34f1d9da179c/test/jit/test_models.py#L402

Another failure is this:
https://app.circleci.com/pipelines/github/pytorch/pytorch/194791/workflows/41dd2318-66e5-416b-8f87-e9fb6e8151e5/jobs/6428574
The code is here:
https://github.com/pytorch/pytorch/blob/ca1b8ebbcb8f959c4044d5832acb34f1d9da179c/test/test_nn.py#L11089

Can you take check your code?

Before doing any update, please rebase your code after the pytorch build status in the following link goes back to all green.
https://ezyang.github.io/pytorch-ci-hud/build2/pytorch-master
If you code can pass all the test, that's fine then, otherwise, there might be something wrong within your code.

**Comment by mingfeima** [comment]
@glaringlee sorry for the delay on this patch, i have been occupied by some other high priority jobs. 
The asan test errors shall be gone after replacing remainder vectorization math with scalar math. Though i thought this might be a false alarm, still it is better to have it fixed.
I am now considering to unify func signature from `_thnn_fused_lstm_cell` of both CPU and CUDA backend, thus the code would be much simplified.
Also I am integrating onednn/mkldnn rnn APIs into pytorch, will compare onednn/mkldnn with this thnn implementation from feature wise and performance wise.

**Comment by glaringlee** [comment]
@mingfeima i will review this later today or next Monday.
Can you fix the lint error at this point? (flake8)
And just personal interest, can you explain a bit why asan test breaks previously? Thanks a lot.

**Comment by glaringlee** [comment]
@mingfeima been busy recently, I went through the modified part, it looks good to me.
Please fix the lint error in your next update, thanks a lot.

**Comment by mingfeima** [comment]
@glaringlee Ohhh... sorry for the delay. My github notifications are stupidly blocked by outlook.

About this patch, I was not so sure whether I want to continue to push it. 
Since the performance improvement is mediocre and too much change in RNN.cpp (due to the double backward support), making the rnn code a little bit too complexed.

Too achieve optimal performance, need to make sure the weight is consistently preserved in cache during the iteration of time sequence. Let's close this one for now.
